core/internal/filetransfer/file_transfer_gcs.go

```diff
@@ -0,0 +1,243 @@
+				return nil
+			}
+			ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error while downloading file for reference %s: %v", reference, err))
+			return err
+		}
+	default:
+		object := bucket.Object(objectName)
+		objAttrs, err := object.Attrs(ft.ctx)
+		if err != nil {
+			// this file is probably a folder, skip to avoid an unnecessary error
+			if isDir(reference, task.Size) {
+				ft.logger.Debug("gcs file transfer: download: skipping reference because it seems to be a folder", "reference", reference)
+				return nil
+			}
+			ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: unable to fetch object attributes for %s for reference %s: %v", objectName, reference, err))
+			return err
+		}
+		if objAttrs.Etag != task.Digest {
+			err := fmt.Errorf("gcs file transfer: download: digest/etag mismatch for reference %s: etag %s does not match expected digest %s", reference, objAttrs.Etag, task.Digest)
+			ft.logger.CaptureError(err)
+			return err
+		}
+		err = ft.DownloadFile(object, task.Path)
+		if err != nil {
+			ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error while downloading file for reference %s: %v", reference, err))
+			return err
+		}
+	}
+	return nil
+}
+
+// DownloadFile downloads the contents of object into a file at path localPath
+func (ft *GCSFileTransfer) DownloadFile(object *storage.ObjectHandle, localPath string) error {
+	r, err := object.NewReader(ft.ctx)
+	if err != nil {
+		ft.logger.CaptureError(fmt.Errorf("unable to create reader for object %s, received error: %v", object.ObjectName(), err))
+		return err
+	}
+	defer r.Close()
+
+	// Check if the directory exists, and create it if it doesn't
+	dir := path.Dir(localPath)
+	if _, err := os.Stat(dir); os.IsNotExist(err) {
+		if err := os.MkdirAll(dir, os.ModePerm); err != nil {
+			return err
+		}
+	} else if err != nil {
+		return err
+	}
+
+	// open the file for writing and defer closing it
+	file, err := os.Create(localPath)
+	if err != nil {
+		return err
+	}
+	defer func(file *os.File) {
+		if err := file.Close(); err != nil {
+			ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error closing file %s: %v", localPath, err))
+		}
+	}(file)
+
+	_, err = io.Copy(file, r)
+	if err != nil {
+		ft.logger.CaptureError(fmt.Errorf("gcs file transfer: download: error copying object %s into file %s: %v", object.ObjectName(), localPath, err))
+		return err
+	}
+
+	return nil
+}
+
+// getDownloadFilePath returns the file path to download the file to, removing duplicate info from the extension
+func getDownloadFilePath(objectName string, prefix string, baseFilePath string) string {
+	ext, _ := strings.CutPrefix(objectName, prefix)
+	localPath := baseFilePath + ext
+	return localPath
+}
+
+// isDir returns true if the reference path and its size indicate that it might be a directory
+func isDir(reference string, size int64) bool {
+	return path.Ext(reference) == "" && size == 0
```

Comment:
This seems flaky to me...to check in terms of 0 size...unsure what would be the best way but have you considered using the GCSClient itself to check if a reference path is a directory? Would that be too slow/expensive?

Reply:
Some digging and AI tells me this might be another option - which uses the attributes of the `ObjectHandle` to see if the object is a "prefix", in which case its a directory. 

```go
func (ft *GCSFileTransfer) isDir(object *storage.ObjectHandle) (bool, error) {
    attrs, err := object.Attrs(ft.ctx)
    if err != nil {
        return false, fmt.Errorf("gcs file transfer: download: error getting object attributes: %v", err) 
    }
    return attrs.Prefix != "", nil 
}
```

Reply:
definitely agree that this is flaky and not an ideal way to check this- this is related to the bug that @moredatarequired and I found earlier this week. originally we wanted to check if the key ended with a '/' and if it has size zero, but it turns out we actually don't store that backslash in the manifest entry. I tried to approximate this as much as I could, but it isn't the best. 

I don't think we can use `object.Attrs` in the suggested way as it errors out unless there is a forward slash at the end of these directory paths, and also the `Prefix` field is still an empty string for these objects when we do add a "/", so we can't use that to distinguish between folders and files.

Something more accurate that we can do is check if the key ends in a backslash (which is the case for the objects we get from `bucket.Objects()`), or if it doesn't have an extension and `object.Attrs` fails with the unaltered key? something like this:

```go
func (ft *GCSFileTransfer) isDir(object *storage.ObjectHandle) (bool, err) {
    if (strings.HasSuffix(object.ObjectName(), "/") && object.Size == 0) {
        return true, nil
    } else if (path.Ext(object.ObjectName()) == "" && size == 0) {
        _, err := object.Attrs(ft.ctx)
        // this should only return this error in the case that its a folder. Otherwise, we wouldn't even be able to get the object 
        if err != nil && errors.Is(err, storage.ErrObjectNotExist) {
            return true, nil
        } else if (err != nil) {
            return nil, err
        }
    }
    return false, nil
}
```

I'm a bit hesitant about doing this because its relying on the package's inconsistencies, so please feel free to push back/suggest other methods!

Reply:
actually, I think I'm mistaken here. `bucket.Object()` will always return an object handle, whether that key exists or not. I'll start a thread in our channel to discuss how to check this